language/move-vm/runtime/src/native_functions.rs

```diff
@@ -1,139 +1,84 @@
-    DestroySigner,
+use std::collections::hash_map;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::fmt::Write;
+use std::hash::Hash;
+
+pub type NativeFunction =
+    fn(&mut NativeContext, Vec<Type>, VecDeque<Value>) -> PartialVMResult<NativeResult>;
+
+pub(crate) struct NativeFunctions(
+    HashMap<AccountAddress, HashMap<String, HashMap<String, NativeFunction>>>,
+);
+
+fn get_mut_or_insert_with<K, V, F>(map: &mut HashMap<K, V>, key: K, mk_value: F) -> &mut V
+where
+    F: FnOnce() -> V,
+    K: Eq + Hash,
+{
+    match map.entry(key) {
+        hash_map::Entry::Occupied(entry) => entry.into_mut(),
+        hash_map::Entry::Vacant(entry) => entry.insert(mk_value()),
+    }
 }
 
-impl NativeFunction {
-    pub(crate) fn resolve(
-        module_address: &AccountAddress,
+impl NativeFunctions {
+    pub fn resolve(
+        &self,
+        addr: &AccountAddress,
         module_name: &str,
-        function_name: &str,
+        func_name: &str,
     ) -> Option<NativeFunction> {
-        use NativeFunction::*;
-
-        let case = (module_address, module_name, function_name);
-        Some(match case {
-            (&CORE_CODE_ADDRESS, "Hash", "sha2_256") => HashSha2_256,
-            (&CORE_CODE_ADDRESS, "Hash", "sha3_256") => HashSha3_256,
-            (&CORE_CODE_ADDRESS, "BCS", "to_bytes") => BCSToBytes,
-            (&CORE_CODE_ADDRESS, "Signature", "ed25519_validate_pubkey") => PubED25519Validate,
-            (&CORE_CODE_ADDRESS, "Signature", "ed25519_verify") => SigED25519Verify,
-            (&CORE_CODE_ADDRESS, "Vector", "length") => VectorLength,
-            (&CORE_CODE_ADDRESS, "Vector", "empty") => VectorEmpty,
-            (&CORE_CODE_ADDRESS, "Vector", "borrow") => VectorBorrow,
-            (&CORE_CODE_ADDRESS, "Vector", "borrow_mut") => VectorBorrowMut,
-            (&CORE_CODE_ADDRESS, "Vector", "push_back") => VectorPushBack,
-            (&CORE_CODE_ADDRESS, "Vector", "pop_back") => VectorPopBack,
-            (&CORE_CODE_ADDRESS, "Vector", "destroy_empty") => VectorDestroyEmpty,
-            (&CORE_CODE_ADDRESS, "Vector", "swap") => VectorSwap,
-            (&CORE_CODE_ADDRESS, "Event", "write_to_event_store") => AccountWriteEvent,
-            (&CORE_CODE_ADDRESS, "DiemAccount", "create_signer") => CreateSigner,
-            (&CORE_CODE_ADDRESS, "Debug", "print") => DebugPrint,
-            (&CORE_CODE_ADDRESS, "Debug", "print_stack_trace") => DebugPrintStackTrace,
-            (&CORE_CODE_ADDRESS, "Signer", "borrow_address") => SignerBorrowAddress,
-            // functions below this line are deprecated and remain only for replaying old transactions
-            (&CORE_CODE_ADDRESS, "DiemAccount", "destroy_signer") => DestroySigner,
-            _ => return None,
-        })
+        self.0
+            .get(addr)
+            .and_then(|modules| modules.get(module_name))
+            .and_then(|funcs| funcs.get(func_name).cloned())
```

Comment:
`.get(addr)?.get(module_name)?.get(func_name).cloned()`

Reply:
Ah, and it can be `.copied()` too

Reply:
Thanks for catching this! For some reason I always subconsciously think that ? only works for Result but not Option...